KAFKA-16686: Wait for given offset in TopicBasedRemoteLogMetadataManagerTest

gaurav-narula opened the pull request:
Some tests in TopicBasedRemoteLogMetadataManagerTest flake because `waitUntilConsumerCatchesUp` may break early before consumer manager has caught up with all the events.

This PR adds an expected offsets for leader/follower metadataOffset partitions and ensures we wait for the offset to be at least equal to the argument to avoid flakyness.

Refer [Gradle Enterprise Report](https://ge.apache.org/scans/tests?search.timeZoneId=Europe%2FLondon&tests.container=org.apache.kafka.server.log.remote.metadata.storage.TopicBasedRemoteLogMetadataManagerTest) for more information on flakyness.

gaurav-narula commented:
CC: @clolov @satishd

kamalcph commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerTest.java:
```diff
@@ -176,17 +178,18 @@ private void waitUntilConsumerCatchesUp(TopicIdPartition newLeaderTopicIdPartiti
             if (leaderMetadataPartition == followerMetadataPartition) {
-                if (topicBasedRlmm().readOffsetForPartition(leaderMetadataPartition).orElse(-1L) >= 1) {
+                Assertions.assertEquals(targetLeaderMetadataPartitionOffset, targetFollowerMetadataPartitionOffset);
+                if (topicBasedRlmm().readOffsetForPartition(leaderMetadataPartition).orElse(-1L) >= targetLeaderMetadataPartitionOffset) {
```
previously, we were waiting for `>=1`, after this change, `>=0`. This will make the test more flaky.

when the leader and follower partitions are mapped to the same partition, then we have to wait for twice the amount of messages.

kamalcph reviewed (commented):
Thanks @gaurav-narula for the patch! Left minor comments to address. 

We can rewrite this test to be concise. I'll file a separate ticket for this.

satishd commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerTest.java:
```diff
@@ -152,15 +152,17 @@ public void testNewPartitionUpdates() throws Exception {
                                             TopicIdPartition newFollowerTopicIdPartition,
-                                            long timeoutMs) throws TimeoutException {
+                                            long timeoutMs,
+                                            long targetLeaderMetadataPartitionOffset,
```
These parameters will not help much here as this method was written for `testNewPartitionUpdates` but other tests in this class used the functionality with the gaps. It is better to relook at those usecases and refactor this method respectively.

satishd reviewed (commented):
Thanks @gaurav-narula for the PR, left a meta comment.

gaurav-narula commented:
Thanks for the feedback @kamalcph @satishd!

I've modified the tests so that we propagate a `Consumer<RemoteLogMetadata>` down to `ConsumerTask` and use it only for tests.

This allows us to replace `waitUntilConsumerCatchesUp` with `TestUtils.waitForCondition` to actually wait on the consumption of all expected `RemoteLogMetadata` objects we set up in the tests instead of relying on offsets which is ambiguous.

Please have a look and let me know your thoughts!

kamalcph commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -153,6 +173,7 @@ public void run() {
 
     private void processConsumerRecord(ConsumerRecord<byte[], byte[]> record) {
         final RemoteLogMetadata remoteLogMetadata = serde.deserialize(record.value());
+        onConsume.accept(remoteLogMetadata);
```
This is not the correct way to capture the events. Assume that the testcase don't want to process an event (`shouldProcess` check returns false). We don't want that event to be captured. 

Instead, we can have a setter method for `RemotePartitionMetadataStore` and pass a custom implementation similar to [DummyEventHandler](https://github.com/apache/kafka/blob/trunk/storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java#L374) where we can capture the event and delegate it to the real implementation.

gaurav-narula replied:
Thanks for the suggestion! This made me realise there's also a possible race where even after `RemotePartitionMetadataStore::handleRemoteLogSegmentMetadata` is invoked, the assertions on `topicBasedRlmm().listRemoteLogSegments` may fail because `remoteLogMetadataCache.isInitialized()` may return false.

Inspired by your suggestion to hook on `RemotePartitionMetadataStore`, I've modified `TopicBasedRemoteLogMetadataManagerHarness` to accept a spy object for it which is passed down to `ConsumerTask`. The tests are modified to ensure `handleRemoteLogSegmentMetadata` and `markInitialized` are invoked appropriate number of times.

kamalcph replied:
Thanks for updating the test, the approach LGTM!

Nit: Why are we using the `supplier` pattern instead of adding a setter to `TopicBasedRemoteLogMetadataManager` and marking it as `visibleForTesting`?

gaurav-narula replied:
> Nit: Why are we using the supplier pattern instead of adding a setter to TopicBasedRemoteLogMetadataManager and marking it as visibleForTesting?

IIUC, you're alluding to something similar we do for `remoteLogMetadataTopicPartitioner` at https://github.com/apache/kafka/blob/8a9dd2beda90f04180e71b406657d9da388d359e/storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerHarness.java#L108.

That can work, but I feel it's very easy to introduce a race inadvertently since `TopicBasedRemoteLogMetadataManager::configure` spawns a thread (https://github.com/apache/kafka/blob/8a9dd2beda90f04180e71b406657d9da388d359e/storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java#L368). In fact, `remoteLogMetadataTopicPartitioner` is prone to a race, where if the test thread yields before line 109 is executed, the `ProducerManager` and `ConsumerManager` instances can get instantiated with incorrect `remoteLogMetdataTopicPartitioner` instance (https://github.com/apache/kafka/blob/8a9dd2beda90f04180e71b406657d9da388d359e/storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java#L421).

We can avoid it by invoking the setter before calling `TopicBasedRemoteLogMetadataManager::configure` but I feel it's easier to enforce it by using a `Supplier` instead. Either way, I feel this race should be fixed as well now :)

gaurav-narula replied:
I've created https://issues.apache.org/jira/browse/KAFKA-16712

gaurav-narula committed:
KAFKA-16686: Improve waiting in TopicBasedRemoteLogMetadataManagerTest
Some tests in TopicBasedRemoteLogMetadataManagerTest flake because
`waitUntilConsumerCatchesUp` may break early before consumer manager has
caught up with all the events.

This change allows passing a spy object for `RemotePartitionMetadataStore`
down to `ConsumerTask` which allows the test code to ensure the methods
on it were invoked appropriate number of times before performing
assertions.

Refer
[Gradle Enterprise Report](https://ge.apache.org/scans/tests?search.timeZoneId=Europe%2FLondon&tests.container=org.apache.kafka.server.log.remote.metadata.storage.TopicBasedRemoteLogMetadataManagerTest)
for more information on flakyness.

kamalcph reviewed (approved):
LGTM, thanks for fixing the flaky test!

chia7712 commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -89,14 +90,16 @@ public class TopicBasedRemoteLogMetadataManager implements RemoteLogMetadataMana
     private volatile boolean initializationFailed;
+    private final Supplier<RemotePartitionMetadataStore> remoteLogMetadataManagerSupplier;
 
     public TopicBasedRemoteLogMetadataManager() {
```
Could you add comments to say this default constructor is required as we create `RemoteLogMetadataManager` dynamically?

gaurav-narula replied:
Sure, addressed both in [a8ba568](https://github.com/apache/kafka/pull/15885/commits/a8ba568bb444d0f0b852c6baf63eb343bd8eee71)

chia7712 commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -89,14 +90,16 @@ public class TopicBasedRemoteLogMetadataManager implements RemoteLogMetadataMana
 
     // Visible for testing.
-    public TopicBasedRemoteLogMetadataManager(boolean startConsumerThread) {
+    public TopicBasedRemoteLogMetadataManager(boolean startConsumerThread, Supplier<RemotePartitionMetadataStore> remoteLogMetadataManagerSupplier) {
```
It seems `package-private` is enough in testing, right?

chia7712 reviewed (commented):
@gaurav-narula nice fix!

gaurav-narula committed:
Address review comments

chia7712 reviewed (approved):
LGTM

I have re-trigger QA. will merge it if no objection.

satishd reviewed (approved):
Thanks @gaurav-narula for addressing the review comments, this approach LGTM.

gaurav-narula commented:
@chia7712 looks like we still suffer from thread leaks in CI :( I've rebased from trunk to trigger CI again

chia7712 commented:
>  looks like we still suffer from thread leaks in CI :( I've rebased from trunk to trigger CI again

I have noticed that too. so sad :(
